lib/ansible/plugins/strategy/__init__.py

```diff
@@ -703,6 +711,8 @@ def _add_host(self, host_info, iterator):
             # Check if host in inventory, add if not
             if host_name not in self._inventory.hosts:
                 self._inventory.add_host(host_name, 'all')
+                self._hosts_cache.append(host_name)
```

Comment:
Is this actually correct? Since you're now driving the play iterator off this cache, you should probably either:
* never add a host to `_hosts_cache` here, since the current play_iterator impl only sets up blocks in __init__, it'll never find blocks/tasks to run anyway, but that could theoretically change, which leads me to ...
* only add the host if the current play would've included it (much harder, but technically more correct)

Reply:
I think the only user-observable issue here right now is that `ansible_play_hosts(_all)` now include the new hosts when they shouldn't, since they aren't executing. The next strategy build will pick up the new hosts if applicable.

Reply:
Hrm, I'm not sure.  Previously we would call `get_hosts` again in many places, and I believe it would have pulled in the new host.  I think this is right.  I'll have to play around with devel and this to see if there is a change, but I believe without this we could miss some feature, such as when populating `ansible_play_hosts_all` where the new host matches the existing pattern.

Reply:
Ok, I see what you were getting at.  Not the manipulation of both, but that only `_hosts_cache` shouldn't be manipulated.  `_hosts_cache_all` should be though.

I compared:

```
when: "'foo' in ansible_play_batch"
```

against this PR and devel, and this PR introduces a change there.